src/http_microhttpd.c

```diff
@@ -589,19 +593,37 @@ static int encode_and_redirect_to_splashpage(struct MHD_Connection *connection,
 		}
 	}
 
-	if (encoded[0]) {
-		safe_asprintf(&splashpageurl, "http://%s:%u/%s?redir=%s",
-			config->gw_address, config->gw_port, config->splashpage, encoded);
+	if (config->fas_port) {
+		safe_asprintf(&ip_address, "%s", config->gw_address);
+
+		if (config->fas_remoteip) {
+			safe_asprintf(&ip_address, "%s", config->fas_remoteip);
+		}
+		safe_asprintf(&splashpageurl, "http://%s:%u%s",
+			ip_address, config->fas_port, config->fas_path);
+		if (config->fas_secure_enabled != 1) {
+			querydl = "&";
+			safe_asprintf(&splashpageurl, "%s?authaction=http://%s:%u/%s/?tok=%s",
+				splashpageurl, config->gw_address, config->gw_port, config->authdir, token);
+		} else {
+			querydl = "&";
+			safe_asprintf(&splashpageurl, "%s?clientip=%s",
+				splashpageurl, token);
+		}
 	} else {
 		safe_asprintf(&splashpageurl, "http://%s:%u/%s",
 			config->gw_address, config->gw_port, config->splashpage);
 	}
 
+	if (encoded[0]) {
+		safe_asprintf(&splashpageurl, "%s%sredir=%s", splashpageurl, querydl, encoded);
```

Comment:
You leak memory here when you use safe_asprintf multiple times on variable splashpageurl.